src/realm/group_writer.cpp

```diff
@@ -301,25 +302,295 @@ GroupWriter::MapWindow* GroupWriter::get_window(ref_type start_ref, size_t size)
+    std::cout << "  Reachable: ";
+    ref_type last = 0;
+    for (auto& i : reachables) {
+        if (i.ref != last) {
+            last = i.ref;
+            std::cout << std::endl << "    " << i.ref << " - " << i.ref + i.size << " : " << i.version;
+        }
+        else {
+            std::cout << ", " << i.version;
+        }
+    }
+    std::cout << std::endl << "  Backdating:";
+#endif
+#endif
+    size_t limit = m_not_free_in_file.size();
+    for (size_t index = 0; index < limit; ++index) {
+        auto& entry = m_not_free_in_file[index];
+#ifdef REALM_DEBUG
+#if REALM_ALLOC_DEBUG
+        std::cout << std::endl
+                  << "    Considering [" << entry.ref << ", " << entry.size << "]-" << entry.released_at_version
+                  << " {";
+#endif
+        unsigned k;
+        for (k = 0; k < reachables.size(); ++k) {
+            if (reachables[k].ref + reachables[k].size > entry.ref)
+                break;
+        }
+        bool referenced = false;
+        while (k < reachables.size() && reachables[k].ref + reachables[k].size > entry.ref &&
+               reachables[k].ref < entry.ref + entry.size) {
+#if REALM_ALLOC_DEBUG
+            std::cout << reachables[k].version << " ";
+#endif
+            ++k;
+            referenced = true;
+        }
+#if REALM_ALLOC_DEBUG
+        if (!referenced) {
+            std::cout << "none";
+        }
+        std::cout << "} ";
+#endif
+#endif
+        // early out if the reference is to the most recent version
+        if (entry.released_at_version == m_current_version) {
+#ifdef REALM_DEBUG
+            REALM_ASSERT(!referenced);
+#endif
+            continue;
+        }
+        // early out for references before oldest freelist:
+        if (entry.released_at_version <= this->m_oldest_reachable_version) {
+#ifdef REALM_DEBUG
+            REALM_ASSERT(!referenced);
+#endif
+            continue;
+        }
+        auto earlier_it = get_earlier(entry.released_at_version);
+        // if no earlier versions than that exists, we're done
+        if (earlier_it == nullptr) {
+#if REALM_ALLOC_DEBUG
+            std::cout << "  backdating [" << entry.ref << ", " << entry.size << "] to zero (no earlier freelist)";
+#endif
+#ifdef REALM_DEBUG
+            REALM_ASSERT(!referenced);
+#endif
+            entry.released_at_version = 0;
+            continue;
+        }
+        // earlier versions exist, we need to examine the freelist for it
+#if REALM_ALLOC_DEBUG
+        std::cout << " - earlier freelist: " << earlier_it->version;
+#endif
+        auto covering_blocks_released_at = find_cover_for(entry, earlier_it);
+        if (covering_blocks_released_at != entry.released_at_version) {
+#ifdef REALM_DEBUG
+            REALM_ASSERT(!referenced);
+#endif
+            entry.released_at_version = covering_blocks_released_at;
+            if (entry.released_at_version) {
+                // retry backdating from the newly found older version:
+                --index;
```

Comment:
I would prefer that we increment the index when we are done with this entry. It is a bit confusing that we decrement the index because we know that it will be incremented just after.

Reply:
I wrote it like that at first. But it implies that we'll need an explicit increment at all the other places where we do "continue". Is that what you prefer?